cpp/src/arrow/compute/row/grouper_benchmark.cc

```diff
@@ -0,0 +1,127 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/util/key_value_metadata.h"
+#include "arrow/util/string.h"
+#include "benchmark/benchmark.h"
+
+#include "arrow/compute/row/grouper.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/testing/random.h"
+#include "arrow/util/benchmark_util.h"
+
+namespace arrow {
+namespace compute {
+
+constexpr auto kSeed = 0x0ff1ce;
+constexpr int64_t kRound = 256;
+
+static ExecBatch MakeRandomExecBatch(const DataTypeVector& types, int64_t num_rows,
+                                     double null_probability,
+                                     int64_t alignment = kDefaultBufferAlignment,
+                                     MemoryPool* memory_pool = nullptr) {
+  random::RandomArrayGenerator rng(kSeed);
+  auto num_types = static_cast<int>(types.size());
+
+  // clang-format off
+  auto metadata = key_value_metadata(
+      {
+        "null_probability",
+        "true_probability",
+        "unique"
+      },
+      {
+          internal::ToChars(null_probability),
+          internal::ToChars(null_probability),                     // for boolean type
+          internal::ToChars(static_cast<int32_t>(num_rows * 0.5))  // for string type
+      });
```

Comment:
So, `int32` and `int64` keys will span the whole magnitude as well? Meaning they will be mostly unique?

Reply:
Yes, we hope that there are enough groups so that we can generate as many comparison operations as possible, which can verify that our underlying comparison performance is stable.